src/ir/auth_logic/auth_logic_ast_traversing_visitor.h

```diff
@@ -0,0 +1,322 @@
+            [this](const datalog::Predicate& pred) {
+              return VariantVisit(pred);
+            },
+            [this](const Attribute& attrib) { return VariantVisit(attrib); },
+            [this](const CanActAs& canActAs) {
+              return VariantVisit(canActAs);
+            }},
+        baseFact.GetValue());
+    Result fold_result = FoldResult(std::move(pre_visit_result),
+                                    std::move(variant_visit_result));
+    return PostVisit(baseFact, std::move(fold_result));
+  }
+
+  Result Visit(CopyConst<IsConst, Fact>& fact) final override {
+    Result pre_visit_result = PreVisit(fact);
+    Result base_fact_result =
+        FoldResult(std::move(pre_visit_result), fact.base_fact().Accept(*this));
+    Result fold_result = common::utils::fold(
+        fact.delegation_chain(), std::move(base_fact_result),
+        [this](Result acc, CopyConst<IsConst, Principal> principal) {
+          return FoldResult(std::move(acc), principal.Accept(*this));
+        });
+    return PostVisit(fact, std::move(fold_result));
+  }
+
+  Result Visit(CopyConst<IsConst, ConditionalAssertion>& conditionalAssertion)
+      final override {
+    Result pre_visit_result = PreVisit(conditionalAssertion);
+    Result lhs_result = FoldResult(std::move(pre_visit_result),
+                                   conditionalAssertion.lhs().Accept(*this));
+    Result fold_result = common::utils::fold(
+        conditionalAssertion.rhs(), std::move(lhs_result),
+        [this](Result acc, CopyConst<IsConst, BaseFact> baseFact) {
+          return FoldResult(std::move(acc), baseFact.Accept(*this));
+        });
+    return PostVisit(conditionalAssertion, std::move(fold_result));
+  }
+
+  Result Visit(CopyConst<IsConst, Assertion>& assertion) final override {
+    Result pre_visit_result = PreVisit(assertion);
+    Result variant_visit_result =
+        std::visit(raksha::utils::overloaded{
+                       [this](const Fact& fact) { return VariantVisit(fact); },
+                       [this](const ConditionalAssertion& condAssertion) {
+                         return VariantVisit(condAssertion);
+                       }},
+                   assertion.GetValue());
+    Result fold_result = FoldResult(std::move(pre_visit_result),
+                                    std::move(variant_visit_result));
+    return PostVisit(assertion, std::move(fold_result));
+  }
+
+  Result Visit(
+      CopyConst<IsConst, SaysAssertion>& saysAssertion) final override {
+    Result pre_visit_result = PreVisit(saysAssertion);
+    Result principal_result = FoldResult(
+        std::move(pre_visit_result), saysAssertion.principal().Accept(*this));
+    Result fold_result = common::utils::fold(
+        saysAssertion.assertions(), std::move(principal_result),
+        [this](Result acc, CopyConst<IsConst, Assertion> assertion) {
+          return FoldResult(std::move(acc), assertion.Accept(*this));
+        });
```

Comment:
I started doing this, but because the body of the function you suggested calls `Accept` on element, the template really needs to specify that Element implements this method. One way to do this would be to write an abstract class that defines the methods all AST nodes have and include `Accept` in this abstract class and then have all the AST nodes extend from that. This seems like a more substantial change that belongs in another PR. So I wrote an issue to track this: #648

Reply:
Templates don't need to specify anything; they'll just try subbing in the type implied by the arguments into the type parameter and see whether the code compiles.

Although, I do now realize that the order of the `Container` and `Element` type parameters should be swapped, as `Container` is inferrable from arguments and `Element` will probably need to be explicitly specified.

Reply:
Ah I mixed up templates with Generics from Java again. Swapping the order must have been it. It's implemented now.